File: src/page/info/filter/EnemyEditBox.java
package page.info.filter;

import page.Page;

import javax.swing.*;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;
import java.util.Vector;

import static utilpc.Interpret.*;

public class EnemyEditBox extends Page {

	private static final long serialVersionUID = 1L;

	private final boolean editable;

	private final Vector<String> vt = new Vector<>();
	private final Vector<String> va = new Vector<>();
	private final AttList trait = new AttList(3, 0);
	private final AttList abis = new AttList(-1, EABIIND.length);
	private final JScrollPane jt = new JScrollPane(trait);
	private final JScrollPane jab = new JScrollPane(abis);

	private boolean changing = false;

	public EnemyEditBox(Page p, boolean edit) {
		super(p);
		editable = edit;
		ini();
	}

	public void setData(int[] vals) {
		changing = true;
		trait.clearSelection();
		for (int i = 0; i < 9; i++)
			if (((vals[0] >> i) & 1) > 0)
				trait.addSelectionInterval(i, i);
		abis.clearSelection();
		for (int i = 0; i < EABIIND.length; i++) {
			int ind = EABIIND[i];
			if (ind < 100 ? ((vals[1] >> ind) & 1) > 0 : ((vals[2] >> (ind - 100 - IMUSFT)) & 1) > 0)
				abis.addSelectionInterval(i, i);
		}
		changing = false;
	}

	@Override
	protected void resized(int x, int y) {
		set(jt, x, y, 0, 0, 200, 400);
		set(jab, x, y, 0, 450, 200, 650);
	}

	private void confirm() {
		int[] ans = new int[3];
		for (int i = 0; i < 9; i++)
			if (trait.isSelectedIndex(i))
				ans[0] |= 1 << i;

		for (int i = 0; i < EABIIND.length; i++)
			if (abis.isSelectedIndex(i))
				if (EABIIND[i] < 100)
					ans[1] |= 1 << EABIIND[i];
				else
					ans[2] |= 1 << (EABIIND[i] - 100 - IMUSFT);
		getFront().callBack(ans);
	}

	private void ini() {
		for (int i = 0; i < 9; i++)
			vt.add(TRAIT[i]);
		for (int i = 0; i < EABIIND.length; i++)
			va.add(EABI[i]);
		trait.setListData(vt);
		abis.setListData(va);
		int m = ListSelectionModel.MULTIPLE_INTERVAL_SELECTION;
		trait.setSelectionMode(m);
		abis.setSelectionMode(m);
		set(trait);
		set(abis);
		add(jt);
		add(jab);
		trait.setEnabled(editable);
		abis.setEnabled(editable);
	}

	private void set(JList<?> jl) {

		jl.addListSelectionListener(new ListSelectionListener() {

			@Override
			public void valueChanged(ListSelectionEvent arg0) {
				if (!changing && !jl.getValueIsAdjusting())
					confirm();
			}

		});
	}

}

File: src/page/info/filter/EnemyEditBox.kt
package page.info.filter

import common.battle.data.DataEntity
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.Page
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import utilpc.Interpret
import java.util.*
import javax.swing.JList
import javax.swing.JScrollPane
import javax.swing.ListSelectionModel
import javax.swing.event.ListSelectionEvent
import javax.swing.event.ListSelectionListener

class EnemyEditBox(p: Page?, private val editable: Boolean) : Page(p) {
    private val vt = Vector<String>()
    private val va = Vector<String>()
    private val trait = AttList(3, 0)
    private val abis = AttList(-1, Interpret.EABIIND.size)
    private val jt: JScrollPane = JScrollPane(trait)
    private val jab: JScrollPane = JScrollPane(abis)
    private var changing = false
    fun setData(vals: IntArray) {
        changing = true
        trait.clearSelection()
        for (i in 0..8) if (vals[0] shr i and 1 > 0) trait.addSelectionInterval(i, i)
        abis.clearSelection()
        for (i in Interpret.EABIIND.indices) {
            val ind: Int = Interpret.EABIIND.get(i)
            if (if (ind < 100) vals[1] shr ind and 1 > 0 else vals[2] shr ind - 100 - Interpret.IMUSFT and 1 > 0) abis.addSelectionInterval(i, i)
        }
        changing = false
    }

    override fun resized(x: Int, y: Int) {
        Page.Companion.set(jt, x, y, 0, 0, 200, 400)
        Page.Companion.set(jab, x, y, 0, 450, 200, 650)
    }

    private fun confirm() {
        val ans = IntArray(3)
        for (i in 0..8) if (trait.isSelectedIndex(i)) ans[0] = ans[0] or (1 shl i)
        for (i in Interpret.EABIIND.indices) if (abis.isSelectedIndex(i)) if (Interpret.EABIIND.get(i) < 100) ans[1] = ans[1] or (1 shl Interpret.EABIIND.get(i)) else ans[2] = ans[2] or (1 shl Interpret.EABIIND.get(i) - 100 - Interpret.IMUSFT)
        front.callBack(ans)
    }

    private fun ini() {
        for (i in 0..8) vt.add(Interpret.TRAIT.get(i))
        for (i in Interpret.EABIIND.indices) va.add(Interpret.EABI.get(i))
        trait.setListData(vt)
        abis.setListData(va)
        val m: Int = ListSelectionModel.MULTIPLE_INTERVAL_SELECTION
        trait.setSelectionMode(m)
        abis.setSelectionMode(m)
        set(trait)
        set(abis)
        add(jt)
        add(jab)
        trait.setEnabled(editable)
        abis.setEnabled(editable)
    }

    private fun set(jl: JList<*>) {
        jl.addListSelectionListener(object : ListSelectionListener {
            override fun valueChanged(arg0: ListSelectionEvent?) {
                if (!changing && !jl.getValueIsAdjusting()) confirm()
            }
        })
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        ini()
    }
}
